api/src/controllers/AggregationProcessorController.js

```diff
@@ -0,0 +1,112 @@
+import catchErrors from 'api/controllers/utils/catchErrors';
+import getOrgFromAuthInfo from 'lib/services/auth/authInfoSelectors/getOrgFromAuthInfo';
+import AggregationProcessor from 'lib/models/aggregationProcessor';
+import { publish } from 'lib/services/queue';
+import sha1 from 'sha1';
+import { get, keys, isObject, filter } from 'lodash';
+import { AGGREGATION_PROCESSOR_QUEUE } from 'lib/constants/aggregationProcessor';
+import getScopeFilter from 'lib/services/auth/filters/getScopeFilter';
+import encode$oid from 'lib/helpers/encode$oid';
+
+export const findOrCreateAggregationProcessor = async ({
+  pipelineString,
+  pipelineHash,
+  windowSize,
+  windowSizeUnits,
+  previousWindowSize,
+  organisation,
+  lrs_id,
+  useWindowOptimization
+}) => {
+  const model = await AggregationProcessor.findOneAndUpdate({
+    organisation,
+    lrs_id,
+    pipelineHash,
+    windowSize,
+    windowSizeUnits,
+    previousWindowSize
+  }, {
+    pipelineString,
+    useWindowOptimization
+  }, {
+    new: true,
+    upsert: true
+  });
+
+  return model;
+};
+
+const canUseWindowOptimization = (pipeline) => {
+  const groupStages = filter(pipeline, value => keys(value)[0] === '$group');
+  if (groupStages.length !== 1) {
+    return false;
+  }
+
+  const group = Object.entries(groupStages[0].$group).find(([key, value]) => {
+    if (
+      isObject(value) &&
+      keys(value).length === 1 &&
+      (
+        keys(value)[0] === '$sum' ||
+        keys(value)[0] === '$first'
+      ) ||
+      key === '_id'
+    ) {
+      return false;
+    }
+    return true;
+  });
+
+  if (group) {
+    return false;
+  }
+
+  return true;
```

Comment:
It seems that this block can be simplified

```suggestion
const group = Object
    .entries(groupStages[0].$group)
    .find(([key, value]) => !(
      isObject(value) &&
      keys(value).length === 1 &&
      (
        keys(value)[0] === '$sum' ||
        keys(value)[0] === '$first'
      ) ||
      key === '_id')
    );

  return !group;
```
```

Reply:
"simplified" ;p